Src/Plugins/Security/SEC101_050.MongoDbConnectionStringValidator.cs

```diff
@@ -0,0 +1,99 @@
+        }
+
+        public static string IsValidStatic(ref string matchedPattern,
+                                           ref Dictionary<string, string> groups,
+                                           ref string failureLevel,
+                                           ref string fingerprint,
+                                           ref string message)
+        {
+            return IsValidStatic(Instance,
+                                 ref matchedPattern,
+                                 ref groups,
+                                 ref failureLevel,
+                                 ref fingerprint,
+                                 ref message);
+        }
+
+        public static string IsValidDynamic(ref string fingerprint, ref string message, ref Dictionary<string, string> options)
+        {
+            return IsValidDynamic(Instance,
+                                  ref fingerprint,
+                                  ref message,
+                                  ref options);
+        }
+
+        protected override string IsValidStaticHelper(ref string matchedPattern,
+                                                      ref Dictionary<string, string> groups,
+                                                      ref string failureLevel,
+                                                      ref string fingerprintText,
+                                                      ref string message)
+        {
+            if (!groups.TryGetNonEmptyValue("account", out string account) ||
+                !groups.TryGetNonEmptyValue("password", out string password) ||
+                !groups.TryGetNonEmptyValue("host", out string host))
+            {
+                return nameof(ValidationState.NoMatch);
+            }
+
+            fingerprintText = new Fingerprint()
+            {
+                Account = account,
+                Password = password,
+                Host = host,
+            }.ToString();
+
+            return nameof(ValidationState.Unknown);
+        }
+
+        protected override string IsValidDynamicHelper(ref string fingerprintText,
+                                                       ref string message,
+                                                       ref Dictionary<string, string> options)
+        {
+            var fingerprint = new Fingerprint(fingerprintText, false);
+
+            string host = fingerprint.Host;
+            string account = fingerprint.Account;
+            string password = fingerprint.Password;
+
+            try
+            {
+                var dbClient = new MongoClient($"mongodb+srv://{account}:{password}@{host}/?connectTimeoutMS=3000");
+                dbClient.ListDatabases();
+            }
+            catch (Exception e)
+            {
+                if (e is MongoAuthenticationException mae)
+                {
+                    if (e.Message.Contains("Unable to authenticate"))
```

Comment:
>Contains [](start = 34, length = 8)

Same note, if possible, prefer Equals comparisons to exact strings. If there's uncertainty (like when we see some strings with a period at the end and others not), StartsWith is useful. Contains is always the most permissive check. Now, you might think that's valuable here and want to make a case for it. I don't know. :)

Reply:
changed to startswith

---
In reply to: [598010958](https://github.com/microsoft/sarif-pattern-matcher/pull/316#discussion_r598010958) [](ancestors = 598010958)